File: 0189_Rotate_Array/solution.cpp
/*
189. Rotate Array
Given an array, rotate the array to the right by k steps, where k is non-negative.

https://leetcode.com/problems/rotate-array/
*/
#include <vector>

using namespace std;

class Solution
{
public:
  void rotate(vector<int> &nums, int k)
  {
    if (nums.empty())
      return;
    k %= nums.size();
    if (k == 0)
      return;
    reverse(begin(nums), end(nums));
    reverse(nums.begin(), nums.begin() + k);
    reverse(begin(nums) + k, end(nums));
  }
};

File: 0189_Rotate_Array/solution.java
package cn.huang.leetcode;

/*
Given an array, rotate the array to the right by k steps, where k is non-negative.

Example 1:

Input: [1,2,3,4,5,6,7] and k = 3
Output: [5,6,7,1,2,3,4]

 */
public class LC_0189_RotateArray {
    public void rotate(int[] nums, int k) {
        int step = k % nums.length;
        rotate(nums, 0, nums.length - 1);
        rotate(nums, 0, step - 1);
        rotate(nums, step, nums.length - 1);

    }


    private void rotate(int[] nums, int start, int end) {
        int left = start;
        int right = end;
        while (left < right) {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }
}
